examples/tokens/ERC1155ownable.vy

```diff
@@ -0,0 +1,358 @@
+    assert fromBalance >= amount, "ERC1155: burn amount exceeds balance"
+    self.balances[id][msg.sender] = fromBalance - amount
+
+    log TransferSingle(operator, msg.sender, ZERO_ADDRESS, id, amount)
+    
+@external
+def burnBatch(ids: DynArray[uint256, BATCH_SIZE], amounts: DynArray[uint256, BATCH_SIZE]):
+    # @dev burn a batch of tokens with the passed IDs
+    # @dev this can be burning non fungible tokens or reducing the balance of existing non-fungible token IDs in the contract
+    # @dev inside the loop ownership will be checked for each token. We can not burn tokens we do not own
+    # @param ids array of ids for the tokens to burn
+    # @param amounts array of amounts of tokens for each ID in the ids array
+    assert not self.paused, "The contract has been paused"
+    assert len(ids) == len(amounts), "ERC1155: ids and amounts length mismatch"
+    operator: address = msg.sender 
+    
+    for i in range(BATCH_SIZE):
+        if i >= len(ids):
+            break
+        assert self.balances[ids[i]][msg.sender] > 0 , "Caller does not own this ID or ZERO balance"
+        id: uint256 = ids[i]
+        amount: uint256 = amounts[i]
+        fromBalance: uint256 = self.balances[id][msg.sender]
+        assert fromBalance >= amount, "ERC1155: burn amount exceeds balance"
+        self.balances[id][msg.sender] = fromBalance - amount
+    
+    log TransferBatch(operator, msg.sender, ZERO_ADDRESS, ids, amounts)
+
+## approval ##
+@external
+def setApprovalForAll(owner: address, operator: address, approved: bool):
+    # @dev set an operator for a certain NFT owner address
+    # @param account the NFT owner address
+    # @param operator the operator address
+    assert owner == msg.sender, "You can only set operators for your own account"
+    assert not self.paused, "The contract has been paused"
+    assert owner != operator, "ERC1155: setting approval status for self"
+    self.operatorApprovals[owner][operator] = approved
+    log ApprovalForAll(owner, operator, approved)
+
+@external 
+@view
+def isApprovedForAll(account: address, operator: address) -> bool:
+    # @dev check wether operator is approved as an operator for the account
+    # @param account the NFT owner address
+    # @param operator the operator address
+    assert self.paused == False, "The contract has been paused"
+    return self.operatorApprovals[account][operator]
+
+@external
+def safeTransferFrom(sender: address, receiver: address, id: uint256, amount: uint256, bytes: bytes32):
+    # @dev transfer token from one address to another.
+    # @param sender the sending account (current owner)
+    # @param receiver the receiving account
+    # @param id the token id that will be sent
+    # @param amount the amount of tokens for the specified id
+    assert not self.paused, "The contract has been paused"
+    assert receiver != ZERO_ADDRESS, "ERC1155: transfer to the zero address"
+    assert receiver != sender, "ERC1155: transfer to self"
```

Comment:
This line isn't necessary per the standard, the worst case scenario here is the sender wastes extra gas
```suggestion
```